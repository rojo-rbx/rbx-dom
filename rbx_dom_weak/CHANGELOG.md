# rbx_dom_weak Changelog

## Unreleased Changes
* Implemented `Default` for `WeakDom`, useful when using Serde or creating an empty `WeakDom`

## 2.6.0 (2023-10-03)
* Added `WeakDom::clone_multiple_into_external` that allows cloning multiple subtrees all at once into a given `WeakDom`, useful for preserving `Ref` properties that point across cloned subtrees ([#364])

[#364]: https://github.com/rojo-rbx/rbx-dom/pull/364

## 2.5.0 (2023-08-09)
* Fix potential stack overflow when creating or inserting into a `WeakDom`. ([#279])
* Added `InstanceBuilder::has_property` for checking if an `InstanceBuilder` defines a given property. ([#283])
* Added `WeakDom::clone_within` and `WeakDom::clone_into_external` for cloning instances into the same or a different `WeakDom`, respectively. ([#312])

[#279]: https://github.com/rojo-rbx/rbx-dom/pull/279
[#283]: https://github.com/rojo-rbx/rbx-dom/pull/283
[#312]: https://github.com/rojo-rbx/rbx-dom/pull/312

## 2.4.0 (2022-06-05)
* Added `WeakDom::into_raw` for enabling fast, non-tree-preserving transformations.
* Added `empty`, `with_class`, and `set_class` methods to `InstanceBuilder`.

## 2.3.0 (2021-10-11)
* Updated to rbx_types 1.3.

## 2.2.0 (2021-07-19)
* Updated to rbx_types 1.2.

## 2.1.0 (2021-07-02)
* Updated to rbx_types 1.1.

## 2.0.0 (2021-06-26) (yanked)
* Updated to rbx_types 1.0.

## 2.0.0-alpha.1 (2021-02-16)
This release is a major, breaking change that introduces many fixes and features.

* `RbxTree` was replaced with `WeakDom`
	* Changed `new` to accept an `InstanceBuilder`, which can contain children.
	* Renamed `get_root_id` to `root_ref`.
	* Renamed `get_instance` to `get_by_ref`.
	* Renamed `get_instance_mut` to `get_by_ref_mut`.
	* Renamed `move_instance` to `transfer`.
	* Renamed `set_parent` to `transfer_within`.
	* Renamed `insert_instance` to `insert`. This method now accepts an `InstanceBuilder`.
	* Removed `remove_instance`, replaced by `destroy` and `transfer`.
	* Removed `iter_all_ids`.
	* Removed `descendants`.
	* Added `root` and `root_mut` for accessing the root instance directly.
* `RbxInstanceProperties` was replaced with `InstanceBuilder`
	* This API is completely different, as it's now a builder. It is now much easier to construct instances.
* `RbxInstance` was replaced with `Instance`
	* This type now directly exposes its values instead of implementing `Deref` for another type.
* Moved types into the `rbx_types` crate, re-exported as `rbx_dom_weak::types`.
	* `RbxId` was replaced with `types::Ref`, which can now represent null referents.
* Added `DomViewer` API from rbx_dom_test to make testing instance trees easier.

## 1.10.1 (2019-12-18)
* Updated `base64`, `md5`, and `uuid` dependencies

## 1.10.0 (2019-09-15)
* Added `RbxTree::set_parent`, for moving instances within a single tree.

## 1.9.0 (2019-07-12)
* Changed `BrickColor` to be much more correct
	* All BrickColor values are now correctly available instead of just palette colors
	* Enum values now align with the `BrickColor.Number` property and `BrickColor.new` constructor
	* `BrickColor` no longer serializes to strings in human-readable Serde formats to avoid information loss. These colors have colliding names:
		* Rust
		* Lilac
		* Gold
		* Deep orange

## 1.8.2 (2019-07-10)
* Added conversion from `Int32` to `BrickColor`. This should fix serialization of `SpawnLocation` instances, which use `int` instead of `BrickColor` in at least the XML model format.

## 1.8.1 (2019-06-20)
* Fixed Serde being able to deserialize `RbxValue` and `RbxValueType`'s internal `__Nonexhaustive` value. This should also fix Serde recommending this value to users in error messages.
* Fixed `BrickColor` serialization to use Roblox color names

## 1.8.0 (2019-06-10)
* Added the `SharedString` type, which is used by Roblox to reduce redundant copies of binary buffers. ([#63](https://github.com/rojo-rbx/rbx-dom/pull/63))

## 1.7.0 (2019-05-14)
* Added conversions from `BrickColor` to `Color3` and `Color3uint8`. ([#46](https://github.com/rojo-rbx/rbx-dom/pull/46))
* Added conversions from `Color3` to `Color3uint8` and vice-versa.
* Changed the human readable serialization of `BinaryString` values to be base64-encoded. This makes JSON-encoded values much smaller. This changes the details of the unspecified serialization of rbx_dom_weak.
* `RbxValueConversion` now derives `PartialEq`.

## 1.6.0 (2019-05-12)
* Added `RbxTree::move_instance` API to move instances from one tree to another.
* Fixed `RbxTree::descendants` to no longer return the instance of the ID you give it. This may break code relying on this (broken) assumption, but was definitely a bug.
* `RbxValueConversion` now derives `Debug` and `Clone` ([#52](https://github.com/rojo-rbx/rbx-dom/issues/52))

## 1.5.0 (2019-05-06)
* Added support for `BrickColor` ([#29](https://github.com/rojo-rbx/rbx-dom/pull/29))
* Added `RbxValue::try_convert_ref`, which can be used to try to coerce values. ([#42](https://github.com/rojo-rbx/rbx-dom/pull/42))
* `Content` values can now be inferred from string literals.
* `RbxValueType` now derives Serde's `Serialize` and `Deserialize` traits

## 1.4.0 (2019-03-27)
* Added support for new types:
	* `NumberRange`
	* [#15](https://github.com/rojo-rbx/rbx-dom/pull/15):
		* `Int64`
		* `Float64`
	* [#20](https://github.com/rojo-rbx/rbx-dom/pull/20):
		* `Rect`
		* `Ray`
		* `ColorSequence`
		* `NumberSequence`

## 1.3.0 (2019-03-14)
* Added `sort_children_by_key` and `sort_children_unstable_by_key` methods to `RbxInstance` to reorder children safely

## 1.2.0 (2019-03-13)
* `RbxTree` and `RbxInstance` are now clonable without an `unimplemented` panic
	* Cloned trees preserve their IDs as-is and there's no public API yet to transplant instances between trees. This is mostly useful for comparing trees before/after a mutation, which we're using in Rojo.

## 1.1.0 (2019-03-11)
* Marked `PhysicalProperties` as `pub`
* Fixed `Serialize` impl for `UnresolvedRbxValue`
	* This removes `Serialize` from `AmbiguousRbxValue`, which should not break any real code.

## 1.0.0 (2019-03-01)
* `RbxValue` and `RbxValueType` can no longer be matched exhaustively, which enables adding new types without breaking code in the future
* Added support for `UDim`, `UDim2`, and `Content` value types
* `PhysicalProperties` is no longer a stub type

## 0.3.1 (2019-02-26)
* Added support for bare bool values when deserializing `UnresolvedRbxValue`
* Implemented `Serialize` for `UnresolvedRbxValue` and `AmbiguousRbxValue`
* Implemented `From<RbxValue>` for `UnresolvedRbxValue`

## 0.3.0 (2019-02-14)
* Renamed crate from `rbx_tree` to `rbx_dom_weak`
* Added support for `Ref` values ([#8](https://github.com/rojo-rbx/rbx-dom/pull/8))
* Added `UnresolvedRbxValue` and `AmbiguousRbxValue`, intended to be used alongside `rbx_reflection` to make specifying values less verbose.

## 0.2.0 (2019-01-25)
* Added new variants for `RbxValue`:
	* Int32
	* Float32
	* Enum
	* Vector2
	* Color3
	* Color3uint8
	* Vector3int16
	* Vector2int16
	* CFrame
	* PhysicalProperties (Stub)

## 0.1.0
* Initial release
* Supports `String`, `Bool`, and `Vector3` property values
