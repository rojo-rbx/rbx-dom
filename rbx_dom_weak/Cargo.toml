[package]
name = "rbx_dom_weak"
description = "Weakly-typed Roblox DOM implementation for Rust"
version = "2.4.0"
license = "MIT"
documentation = "https://docs.rs/rbx_dom_weak"
homepage = "https://github.com/rojo-rbx/rbx-dom"
repository = "https://github.com/rojo-rbx/rbx-dom.git"
readme = "README.md"
authors = ["Lucien Greathouse <me@lpghatguy.com>"]
edition = "2018"

[dependencies]
rbx_types = { version = "1.4.0", path = "../rbx_types", features = ["serde"] }

serde = "1.0.137"

[dev-dependencies]
insta = { version="1.14.1", features=["yaml"] }
