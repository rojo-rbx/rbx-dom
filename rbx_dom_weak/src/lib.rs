//! rbx_dom_weak is a common representation of the Roblox DOM for Rust. It's
//! designed to play nicely with the borrow checker and allows accessing
//! instances by ID in constant time.
//!
//! Constructing a new tree of instances is accomplished by first creating an
//! [`InstanceBuilder`] object that describes a tree of instances and then
//! wrapping it with an [`WeakDom`]:
//!
//! ```
//! use rbx_dom_weak::{InstanceBuilder, WeakDom};
//!
//! let dm = InstanceBuilder::new("DataModel");
//!
//! let mut dom = WeakDom::new(dm);
//!
//! println!("ID of DOM root is {:?}", dom.root_ref());
//! ```
//!
//! Once we have a tree, we can use [`WeakDom::insert`] and
//! [`WeakDom::get_by_ref`] to add instances to the tree and retrieve them.
//!
//! ```
//! use rbx_dom_weak::{InstanceBuilder, WeakDom};
//!
//! let mut dom = WeakDom::new(InstanceBuilder::new("DataModel"));
//!
//! // We can define properties using any type that can be converted to an
//! // rbx_dom_weak::types::Variant.
//! let http_service = InstanceBuilder::new("HttpService")
//!     .with_property("HttpEnabled", true);
//!
//! let http_service_id = dom.insert(dom.root_ref(), http_service);
//!
//! println!("HttpService has ID {:?}", http_service_id);
//! ```
//!
//! To change properties on an instance that's already present in the tree, use
//! [`WeakDom::get_by_ref_mut`]. Note that it isn't possible to add or remove
//! children through this method, use [`WeakDom::insert`] and
//! [`WeakDom::destroy`] instead.

#![deny(missing_docs)]

mod dom;
mod instance;
mod viewer;

pub use rbx_types as types;

pub use ahash::AHashMap;

pub use crate::{
    dom::WeakDom,
    instance::{Instance, InstanceBuilder},
    viewer::{DomViewer, ViewedInstance},
};

/// Helper trait that provides convenience methods for `AHashMap`.
pub trait HashMapExt {
    /// Constructs an empty map.
    fn new() -> Self;

    /// Constructs an empty map with at least the specified capacity.
    fn with_capacity(capacity: usize) -> Self;
}

impl<K, V> HashMapExt for AHashMap<K, V> {
    /// Creates an empty `AHashMap` using the default value for its hasher.
    fn new() -> Self {
        AHashMap::default()
    }

    /// Creates an empty `AHashMap` with at least the specified capacity using
    /// the default value for its hasher.
    fn with_capacity(capacity: usize) -> Self {
        AHashMap::with_capacity_and_hasher(capacity, Default::default())
    }
}
