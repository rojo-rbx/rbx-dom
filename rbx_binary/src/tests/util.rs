use std::{fs, path::Path};

use rbx_dom_weak::DomViewer;

use crate::{
    deserializer::{DecodeOptions, DecompressedFile},
    text_deserializer::DecodedModel,
    to_writer,
};

/// Run a basic gauntlet of tests to verify that the serializer and deserializer
/// can handle this model correctly.
pub fn run_model_base_suite(model_path: impl AsRef<Path>) {
    let model_path = model_path.as_ref();

    // The useful name of the test is the folder containing it.
    let model_stem = model_path
        .parent()
        .unwrap()
        .file_stem()
        .unwrap()
        .to_str()
        .unwrap();

    let contents = fs::read(model_path).unwrap();

    // Write out a text version of the test file. This helps when debugging what
    // the actual test file is and also guards us against the test file
    // changing.
    let text_decoded = DecodedModel::from_reader(contents.as_slice());
    insta::assert_yaml_snapshot!(format!("{}__input", model_stem), text_decoded);

    // Decode the test file and snapshot a stable version of the resulting tree.
    // This should properly test the deserializer.
    let file = DecompressedFile::from_reader(contents.as_slice()).unwrap();
    let decoded = file.deserialize(DecodeOptions::read_unknown()).unwrap();
    let decoded_viewed = DomViewer::new().view_children(&decoded);
    insta::assert_yaml_snapshot!(format!("{}__decoded", model_stem), decoded_viewed);

    // Re-encode the model that we decoded. We can't snapshot this directly...
    let decoded_root = decoded.root();
    let top_level_ids = decoded_root.children();
    let mut encoded = Vec::new();
    to_writer(&mut encoded, &decoded, top_level_ids).unwrap();

    // ...but we can snapshot the text representation of what we encoded! In an
    // ideal world, this would be very similar or the same as the text
    // representation of the original test file. In practice, we'll differ
    // slightly in chunk ordering, compression, etc.
    let text_roundtrip = DecodedModel::from_reader(encoded.as_slice());
    insta::assert_yaml_snapshot!(format!("{}__encoded", model_stem), text_roundtrip);

    // As a sanity check, make sure we can decode the re-encoded version of the
    // file.
    //
    // We don't make any assertions about the result right now, as our format
    // support is still lacking. In the future, we should assert that this is
    // the same as the original decoding of the test file.
    DecompressedFile::from_reader(encoded.as_slice())
        .unwrap()
        .deserialize(DecodeOptions::read_unknown())
        .unwrap();
}
