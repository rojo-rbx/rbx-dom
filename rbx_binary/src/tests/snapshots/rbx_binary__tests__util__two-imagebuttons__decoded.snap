---
source: rbx_binary/src/tests/util.rs
expression: decoded_viewed

---
- referent: referent-0
  name: ImageButton
  class: ImageButton
  properties:
    Active:
      Bool: true
    AnchorPoint:
      Vector2:
        - 0.0
        - 0.0
    AttributesSerialize:
      BinaryString: ""
    AutoButtonColor:
      Bool: true
    AutoLocalize:
      Bool: true
    AutomaticSize:
      Enum: 0
    BackgroundColor3:
      Color3:
        - 1.0
        - 1.0
        - 1.0
    BackgroundTransparency:
      Float32: 0.0
    BorderColor3:
      Color3:
        - 0.10588236153125763
        - 0.16470588743686677
        - 0.20784315466880799
    BorderMode:
      Enum: 0
    BorderSizePixel:
      Int32: 1
    ClipsDescendants:
      Bool: false
    Draggable:
      Bool: false
    HoverImage:
      Content: ""
    Image:
      Content: "rbxasset://textures/ui/GuiImagePlaceholder.png"
    ImageColor3:
      Color3:
        - 1.0
        - 1.0
        - 1.0
    ImageRectOffset:
      Vector2:
        - 0.0
        - 0.0
    ImageRectSize:
      Vector2:
        - 0.0
        - 0.0
    ImageTransparency:
      Float32: 0.0
    LayoutOrder:
      Int32: 0
    Modal:
      Bool: false
    NextSelectionDown: "null"
    NextSelectionLeft: "null"
    NextSelectionRight: "null"
    NextSelectionUp: "null"
    Position:
      UDim2:
        - - 0.0
          - 0
        - - 0.0
          - 0
    PressedImage:
      Content: ""
    RootLocalizationTable: "null"
    Rotation:
      Float32: 0.0
    ScaleType:
      Enum: 1
    Selectable:
      Bool: true
    Selected:
      Bool: false
    SelectionImageObject: "null"
    Size:
      UDim2:
        - - 0.0
          - 100
        - - 0.0
          - 100
    SizeConstraint:
      Enum: 0
    SliceCenter:
      Rect:
        - - -1.0
          - -10.0
        - - 8.0
          - 9.0
    SliceScale:
      Float32: 1.0
    SourceAssetId:
      Int64: -1
    Style:
      Enum: 0
    Tags:
      BinaryString: ""
    TileSize:
      UDim2:
        - - 1.0
          - 0
        - - 1.0
          - 0
    Visible:
      Bool: true
    ZIndex:
      Int32: 1
  children: []
- referent: referent-1
  name: ImageButton
  class: ImageButton
  properties:
    Active:
      Bool: true
    AnchorPoint:
      Vector2:
        - 0.0
        - 0.0
    AttributesSerialize:
      BinaryString: ""
    AutoButtonColor:
      Bool: true
    AutoLocalize:
      Bool: true
    AutomaticSize:
      Enum: 0
    BackgroundColor3:
      Color3:
        - 1.0
        - 1.0
        - 1.0
    BackgroundTransparency:
      Float32: 0.0
    BorderColor3:
      Color3:
        - 0.10588236153125763
        - 0.16470588743686677
        - 0.20784315466880799
    BorderMode:
      Enum: 0
    BorderSizePixel:
      Int32: 1
    ClipsDescendants:
      Bool: false
    Draggable:
      Bool: false
    HoverImage:
      Content: ""
    Image:
      Content: "rbxasset://textures/ui/GuiImagePlaceholder.png"
    ImageColor3:
      Color3:
        - 1.0
        - 1.0
        - 1.0
    ImageRectOffset:
      Vector2:
        - 0.0
        - 0.0
    ImageRectSize:
      Vector2:
        - 0.0
        - 0.0
    ImageTransparency:
      Float32: 0.0
    LayoutOrder:
      Int32: 0
    Modal:
      Bool: false
    NextSelectionDown: "null"
    NextSelectionLeft: "null"
    NextSelectionRight: "null"
    NextSelectionUp: "null"
    Position:
      UDim2:
        - - 0.0
          - 0
        - - 0.0
          - 0
    PressedImage:
      Content: ""
    RootLocalizationTable: "null"
    Rotation:
      Float32: 0.0
    ScaleType:
      Enum: 1
    Selectable:
      Bool: true
    Selected:
      Bool: false
    SelectionImageObject: "null"
    Size:
      UDim2:
        - - 0.0
          - 100
        - - 0.0
          - 100
    SizeConstraint:
      Enum: 0
    SliceCenter:
      Rect:
        - - 0.0
          - 1.0
        - - 5.0
          - 6.0
    SliceScale:
      Float32: 1.0
    SourceAssetId:
      Int64: -1
    Style:
      Enum: 0
    Tags:
      BinaryString: ""
    TileSize:
      UDim2:
        - - 1.0
          - 0
        - - 1.0
          - 0
    Visible:
      Bool: true
    ZIndex:
      Int32: 1
  children: []

